### ISSUE TITLE
Meta issue about unclear error messages

### ISSUE THREAD

**Issue by giordano** [JuliaLang/Pkg.jl#2778]
I feel like Pkg has a general problem with unclear error messages.  I figured I'd open a meta issue to discuss this problem to address it at a more general level, rather than trying to solve individual bugs.

By going through the list of issues I found (before getting tired):

* #1659
* #1714
* #1818
* #1821
* #1853
* #1935
* #1956
* #2097
* #2191
* #2214
* #2308
* #2381
* #2406
* #2476
* #2481
* #2513
* #2528
* #2554
* #2560
* #2588
* #2602
* #2672
* #2697
* #2701
* #2738
* #2748
* #2774
* #2776

Some issues are already closed, that's not the point, I just wanted to show that very often users, including experienced ones, face error messages that are unactionable, confusing or also completely misleading.

I totally understand that these things can go wrong in uncountable many ways and it's impossible to cater for all possible situations, in fact many times the problems are on the user's side, rather than Pkg's, but users have no idea of what they did wrong or how to fix their issue.  They are only left with unclear messages and no clues of what to do.

For example, when there is something wrong in an environment, it'd be useful to explicitly point out _which_ is the fault environment.

**Comment by IanButterworth** [comment]
I agree. It'd be good to do a push to fix the open issues, and I wonder if the error handling tooling could be bolstered to make it easier. For instance make `pkgerror` accept some standard kwargs like a project/manifest filepath, a sourcefile, a cachefile, etc. that print in a fixed way, to make errors more consistent and easier for Pkg devs to make richer errors?

**Comment by IanButterworth** [comment]
Perhaps a GitHub project instead?

**Comment by giordano** [comment]
> #2776 this is not an error message problem, it is a bug.

The user created by mistake a circular dependency that shouldn't have been there in the first place.  Pkg knows that there is a circular dependency but it doesn't tell you (or not in all cases).

> Same with #2214.

I'm pretty sure #2214 is the same as #2097, so that one can be closed as well.  The solution was to rework the manifest to provide...more informative error messages and let the users know that they're using incompatible manifests.

> What's the point of a meta issue? When do we close this, when all the linked issues are closed? I don't see how this helps get any work done. It comes off as mostly a gripe.

It is a gripe if you want.  I had solved myself issues like #1853 and #2776 before opening the tickets, I could have just moved on, but try to think about how frustrated can be users who are not willing to spend time debugging these issues.

I wanted to show this is a general problem, it isn't specific to a single failure.  The list of issues isn't the list of issues that have to be closed, but it's there to show that users just keep running into different errors and don't have a clue of what to do.  Maybe general error handling/reporting should be reconsidered, instead of (well, in addition to) looking to individual bugs.

**Comment by KristofferC** [comment]
> The user created by mistake a circular dependency that shouldn't have been there in the first place

I don't see why Pkg should not handle it? I've fixed other problems in Pkg where it has bugged on circular dependencies.

> I wanted to show this is a general problem, it isn't specific to a single failure

I mean, it's like opening an issue on Base julia and saying that compilation is slow. OK, now what? 

> Maybe general error handling/reporting should be reconsidered,

PRs implementing some more structured error handling in Julia (or just Pkg) that improves the experience for users are welcome as well as targeted issues/PRs that aim to improve individual error messages. 

But again, I don't see how this issue does anything constructively unless you think that the people that regularly work on Pkg has some notion that the error messages are perfect and they just need to be informed of their badness to realize they should prioritise differently.

**Comment by DilumAluthge** [comment]
How about we convert this issue to a discussion? Then we don't need to determine the criteria for closing.

**Comment by IanButterworth** [comment]
Converting this to a discussion seems like a good idea. I'll do that.

I should add that before this issue, @giordano and I were discussing some of these pkg error issues/gripes and I made a suggestion to make this meta issue to prompt some ideas on how to handle things in general.

On a related note, I recently took a pass through all of the open Pkg issues and was struck by how many diverse ideas and open bugs there are (I closed anything that I considered resolved). I'm wondering if there would be merit to organizing them into some targeted projects, perhaps including one for improving error messaging? - Projects seem to be disabled on this repo, someone with admin would need to enable them
